polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -2616,91 +2633,215 @@ ValueCategory MLIRScanner::EmitBinRem(const BinOpInfo &Info) {
   }
 }
 
+/// Casts index of subindex operation conditionally.
+static Optional<Value> castSubIndexOpIndex(OpBuilder &Builder, Location Loc,
+                                           ValueCategory Pointer,
+                                           ValueRange IdxList, bool IsSigned) {
+  if (Pointer.val.getType().isa<MemRefType>()) {
+    assert(IdxList.size() == 1 && "SubIndexOP accepts just an index");
+    return ValueCategory(IdxList.front(), false)
+        .IntCast(Builder, Loc, Builder.getIndexType(), IsSigned)
+        .val;
+  }
+  return llvm::None;
+}
+
+ValueCategory MLIRScanner::EmitCheckedInBoundsGEP(mlir::Type ElemTy,
+                                                  ValueCategory Pointer,
+                                                  ValueRange IdxList,
+                                                  bool IsSigned, bool) {
+  assert(mlirclang::isPointerOrMemRefTy(Pointer.val.getType()) &&
+         "Expecting pointer or MemRef");
+  assert(std::all_of(IdxList.begin(), IdxList.end(),
+                     [](mlir::Value Val) {
+                       return Val.getType().isa<IntegerType>();
```

Comment:
IntegerType or IndexType?

Reply:
These should be of `IntegerType`, as we will only perform the casting to `IndexType` if we're performing a `SubIndexOp`.